README.md
#Hacker News Panel Indicator
A panel indicator which brings the latest feed from [Hacker News](https://news.ycombinator.com/) right to your desktop. Click on the news feed item to read about it in detail in your web broswer. One can change the duration after which the news feed gets refreshed by adjusting the refresh duration right from the panel.

#Usage

```
git clone https://github.com/mayank26saxena/hn-indicator/

cd hn-indicator

python hn &
```

#PPA
```
sudo add-apt-repository ppa:mayank26saxena/hn-indicator

sudo apt-get update

sudo apt-get install hn-indicator
```

#Run from Dash
![alt tag](https://github.com/mayank26saxena/reddit-indicator/blob/master/screenshots/screenshot3.png)

##Top 10 latest posts from the Hacker News website. 
![alt tag](https://github.com/mayank26saxena/hn-indicator/blob/master/screenshots/screenshot-1.png)

##Adjusting refresh duration.
![alt tag](https://github.com/mayank26saxena/hn-indicator/blob/master/screenshots/screenshot-2.png)

#Contributing
Use github's pull request/issues feature for all contributions.


debian/changelog
hn-indicator (1.4) trusty; urgency=low

  * Fifth Release.

 -- Mayank Saxena <mayank26saxena@gmail.com>  Wed, 15 Jun 2016 01:48:22 +0530
 
hn-indicator (1.3) trusty; urgency=low

  * Fourth Release.

 -- Mayank Saxena <mayank26saxena@gmail.com>  Wed, 15 Jun 2016 00:10:22 +0530
 
 hn-indicator (1.2) trusty; urgency=low

  * Third Release.

 -- Mayank Saxena <mayank26saxena@gmail.com>  Tues, 14 Jun 2016 23:45:07 +0530
 
 hn-indicator (1.1) trusty; urgency=low

  * Second Release.

 -- Mayank Saxena <mayank26saxena@gmail.com>  Tues, 07 Jun 2016 20:38:39 +0530
 
 hn-indicator (1.0) trusty; urgency=low

  * Initial Release.

 -- Mayank Saxena <mayank26saxena@gmail.com>  Sat, 04 Jun 2016 15:28:19 +0530


hn_news.py
import os
import signal
import json
import webbrowser
import pygtk
import gtk
from urllib2 import Request, urlopen, URLError
import appindicator 
import pynotify
from hackernews import HackerNews
import glib


#KEYWORDS
APPINDICATOR_ID = 'hn-indicator'
PYNOTIFY_ID = 'hn-indicator'
LIMIT = 10
ONE_MINUTE = 1000*60
ONE_HOUR = 60
DURATION = 1
REFRESH_DURATION = ONE_MINUTE*ONE_HOUR*DURATION #Feed will get updated every hour as default
SPACE = ' '
DASH = '-'
POINTS = 'Points'

ids = []
titles = []
points = []
urls = []

news = ['?']*LIMIT
item = ['?']*LIMIT

def main():
    indicator = appindicator.Indicator(APPINDICATOR_ID, os.path.abspath('hackernews.png'), appindicator.CATEGORY_COMMUNICATIONS)
    indicator.set_status(appindicator.STATUS_ACTIVE)
    indicator.set_menu(build_menu())
    pynotify.init(APPINDICATOR_ID)
    gtk.main()
    timeout = glib.timeout_add_seconds(60, update_timeout)
    
def build_menu():
    menu = gtk.Menu()

    get_data()

    global LIMIT
    print 'In build_menu() method and value of limit is - ' + str(LIMIT)
    news = ['?']*LIMIT
    item = ['?']*LIMIT
    
    for i in range(0,LIMIT):
        news[i] = str(titles[i]) + SPACE + DASH + SPACE + str(points[i]) + SPACE + POINTS
    
    for i in range(0,LIMIT):    
        item[i] = gtk.MenuItem(news[i])
        item[i].connect('activate', open_url,urls[i])
        menu.append(item[i])

    separator = gtk.SeparatorMenuItem()
    menu.append(separator)

    item_refresh_control = gtk.MenuItem("Adjust refresh time")
    menu_refresh_control = gtk.Menu()
    menu_refresh_control_items = []
    group = [None]
    time_interval_list = [1, 2, 3, 4]
    for i in time_interval_list:
        subitem = gtk.RadioMenuItem(group[0], str(i) + " hours")
        subitem.connect('activate', update_time_interval)
        menu_refresh_control.append(subitem)
        menu_refresh_control_items.append(subitem)
        group = subitem.get_group()
    item_refresh_control.set_submenu(menu_refresh_control)
    menu.append(item_refresh_control)

    separator2 = gtk.SeparatorMenuItem()
    menu.append(separator2)

    item_refresh = gtk.MenuItem('Refresh')
    item_refresh.connect('activate', refresh)
    menu.append(item_refresh)

    item_quit = gtk.MenuItem('Quit')
    item_quit.connect('activate', quit)
    menu.append(item_quit)
            
    menu.show_all()

    global REFRESH_DURATION
    timeout = glib.timeout_add_seconds(REFRESH_DURATION, update_timeout)
    
    return menu


def update_time_interval(source):
    new_update_interval = int(source.get_label().split()[0])
    global REFRESH_DURATION
    global DURATION
    DURATION = new_update_interval
    REFRESH_DURATION = ONE_MINUTE*ONE_HOUR*DURATION
    show_notification(None, "Settings updated!", "Successfully changed the refresh duration settings.")

    
def quit(widget):
    show_notification(None, "Quitting", "Exiting from hn-indicator.")
    gtk.main_quit()

def get_data():
    hn = HackerNews()
    ids = hn.top_stories(limit=LIMIT)
    for i in range(0,LIMIT):
        b = hn.get_item(ids[i])
        titles.append(b.title.encode('utf8'))
        points.append(b.score)
        urls.append(b.url) 

    #print titles
    #print points
    #print urls

def open_url(widget,url):
    if url is not None:
        webbrowser.open(url)
    else:
        show_notification(None, "Error." , "Cannot open URL in browser.")

def show_notification(self, title="Title", msg="Message", timeout=1000):
		notification = pynotify.Notification(title, msg)
		notification.set_timeout(timeout)
		notification.show()

def update_timeout():
    print 'in update_timeout function'
    refresh()
    return True

def refresh():
    ids[:] = []
    titles[:] = []
    points[:] = []  
    urls[:] = []

    build_menu()
    show_notification(None,"Hacker News Widget updated!", "Click on the widget to view updated news feed.")


if __name__ == "__main__" :
    signal.signal(signal.SIGINT, signal.SIG_DFL)
    main()


indicator.py
import os
import signal
import json

from urllib2 import Request, urlopen, URLError

from gi.repository import Gtk as gtk
from gi.repository import AppIndicator3 as appindicator
from gi.repository import Notify as notify


APPINDICATOR_ID = 'score-indicator'

def main():
    indicator = appindicator.Indicator.new(APPINDICATOR_ID, os.path.abspath('sample_icon.svg'), appindicator.IndicatorCategory.SYSTEM_SERVICES)
    indicator.set_status(appindicator.IndicatorStatus.ACTIVE)
    indicator.set_menu(build_menu())
    notify.init(APPINDICATOR_ID)
    gtk.main()

def build_menu():
    menu = gtk.Menu()
    item_joke = gtk.MenuItem('Joke')
    item_joke.connect('activate', joke)
    menu.append(item_joke)
    item_quit = gtk.MenuItem('Quit')
    item_quit.connect('activate', quit)
    menu.append(item_quit)
    menu.show_all()
    return menu

def fetch_joke():
    request = Request('http://api.icndb.com/jokes/random?limitTo=[nerdy]')
    response = urlopen(request)
    joke = json.loads(response.read())['value']['joke']
    return joke

def joke(_):
    notify.Notification.new("<b>Joke</b>", fetch_joke(), None).show()

def quit(_):
    notify.uninit()
    gtk.main_quit()

if __name__ == "__main__":
    signal.signal(signal.SIGINT, signal.SIG_DFL)
    main()


hn.py
import os
import signal
import json
import webbrowser
#import pygtk
import gtk
from urllib2 import Request, urlopen, URLError
import appindicator 
import pynotify
from hackernews import HackerNews
import glib


#KEYWORDS
APPINDICATOR_ID = 'hn-indicator'
PYNOTIFY_ID = 'hn-indicator'
LIMIT = 10
ONE_MINUTE = 1000*60
ONE_HOUR = 60
DURATION = 1
REFRESH_DURATION = ONE_MINUTE*ONE_HOUR*DURATION #Feed will get updated every hour as default
SPACE = ' '
DASH = '-'
POINTS = 'Points'

ids = []
titles = []
points = []
urls = []

news = ['?']*LIMIT
item = ['?']*LIMIT

def main():
    indicator = appindicator.Indicator(APPINDICATOR_ID, os.path.abspath('hackernews.png'), appindicator.CATEGORY_COMMUNICATIONS)
    indicator.set_status(appindicator.STATUS_ACTIVE)
    indicator.set_menu(build_menu())
    pynotify.init(APPINDICATOR_ID)
    gtk.main()
    timeout = glib.timeout_add_seconds(60, update_timeout)
    
def build_menu():
    menu = gtk.Menu()

    get_data()

    global LIMIT
    print 'In build_menu() method and value of limit is - ' + str(LIMIT)
    news = ['?']*LIMIT
    item = ['?']*LIMIT
    
    for i in range(0,LIMIT):
        news[i] = str(titles[i]) + SPACE + DASH + SPACE + str(points[i]) + SPACE + POINTS
    
    for i in range(0,LIMIT):    
        item[i] = gtk.MenuItem(news[i])
        item[i].connect('activate', open_url,urls[i])
        menu.append(item[i])

    separator = gtk.SeparatorMenuItem()
    menu.append(separator)

    item_refresh_control = gtk.MenuItem("Adjust refresh time")
    menu_refresh_control = gtk.Menu()
    menu_refresh_control_items = []
    group = [None]
    time_interval_list = [1, 2, 3, 4]
    for i in time_interval_list:
        subitem = gtk.RadioMenuItem(group[0], str(i) + " hours")
        subitem.connect('activate', update_time_interval)
        menu_refresh_control.append(subitem)
        menu_refresh_control_items.append(subitem)
        group = subitem.get_group()
    item_refresh_control.set_submenu(menu_refresh_control)
    menu.append(item_refresh_control)

    separator2 = gtk.SeparatorMenuItem()
    menu.append(separator2)

    item_refresh = gtk.MenuItem('Refresh')
    item_refresh.connect('activate', refresh)
    menu.append(item_refresh)

    item_quit = gtk.MenuItem('Quit')
    item_quit.connect('activate', quit)
    menu.append(item_quit)
            
    menu.show_all()

    global REFRESH_DURATION
    timeout = glib.timeout_add_seconds(REFRESH_DURATION, update_timeout)
    
    return menu


def update_time_interval(source):
    new_update_interval = int(source.get_label().split()[0])
    global REFRESH_DURATION
    global DURATION
    DURATION = new_update_interval
    REFRESH_DURATION = ONE_MINUTE*ONE_HOUR*DURATION
    show_notification(None, "Settings updated!", "Successfully changed the refresh duration settings.")

    
def quit(widget):
    show_notification(None, "Quitting", "Exiting from hn-indicator.")
    gtk.main_quit()

def get_data():
    hn = HackerNews()
    ids = hn.top_stories(limit=LIMIT)
    for i in range(0,LIMIT):
        b = hn.get_item(ids[i])
        titles.append(b.title.encode('utf8'))
        points.append(b.score)
        urls.append(b.url) 

    #print titles
    #print points
    #print urls

def open_url(widget,url):
    if url is not None:
        webbrowser.open(url)
    else:
        show_notification(None, "Error." , "Cannot open URL in browser.")

def show_notification(self, title="Title", msg="Message", timeout=1000):
		notification = pynotify.Notification(title, msg)
		notification.set_timeout(timeout)
		notification.show()

def update_timeout():
    print 'in update_timeout function'
    refresh()
    return True

def refresh():
    ids[:] = []
    titles[:] = []
    points[:] = []  
    urls[:] = []

    build_menu()
    show_notification(None,"Hacker News Widget updated!", "Click on the widget to view updated news feed.")


if __name__ == "__main__" :
    signal.signal(signal.SIGINT, signal.SIG_DFL)
    main()
